--- spdlog-1.11.0/include/spdlog/common.h	2022-11-02 22:13:08.000000000 +0100
+++ spdlog/include/spdlog/common.h	2023-02-03 11:20:48.986307242 +0100
@@ -165,7 +165,7 @@
 template<class T, class Char = char>
 struct is_convertible_to_basic_format_string
     : std::integral_constant<bool,
-          std::is_convertible<T, fmt::basic_string_view<Char>>::value || std::is_same<remove_cvref_t<T>, fmt::basic_runtime<Char>>::value>
+          std::is_convertible<T, fmt::basic_string_view<Char>>::value || std::is_same<remove_cvref_t<T>, fmt::runtime_format_string<Char>>::value>
 {};
 
 #    if defined(SPDLOG_WCHAR_FILENAMES) || defined(SPDLOG_WCHAR_TO_UTF8_SUPPORT)
